File: 2. Average of Subarrays of Size K.py
"""
Average of Subarrays of Size K
"""
def averages(arr, k):
    n = len(arr)
    if k>n: return []
    window_sum = sum(arr[:k])
    res = [window_sum / k]
    for i in range(k, n):
        window_sum += arr[i] - arr[i-k]
        res.append(window_sum / k)
    return res

def main():
    arr = [1,3,2,6,-1,4,1,8,2]
    print(averages(arr,5))

if __name__ == '__main__':
    main()

# Time Complexity: O(n)
# Space Complexity: O(n) for result

File: 2. Average of Subarrays of Size K.cpp
// Topic: Average of Subarrays of Size K
#include <bits/stdc++.h>
using namespace std;

int main()
{
    vector<int> arr = {1, 3, 2, 6, -1, 4, 1, 8, 2};
    int k = 5;
    int n = arr.size();
    double window_sum = 0;
    for (int i = 0; i < k; i++) window_sum += arr[i];
    cout << window_sum / k << '\n';
    for (int i = k; i < n; i++)
    {
        window_sum += arr[i] - arr[i - k];
        cout << window_sum / k << '\n';
    }
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
